### PR TITLE
Move xds classes for Stubby to xds.client package

### PR DESCRIPTION
**Pull request by larry-safran**
To allow the Stubby team to use XdsClient, moved the basic set of classes to a new package io.grpc.xds.client.  Removed from those classes references to grpc specific things.  Created some interfaces and some grpc spefic extensions of those classes.

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/client/XdsLogger.java [hunk 1]
```diff
@@ -14,9 +14,10 @@
  * limitations under the License.
  */
 
-package io.grpc.xds;
+package io.grpc.xds.client;
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/XdsLogger.java:17]: We have documentation and code referencing this class to enable logging. We're going to need to do this carefully or we tweak how we use it.

File: xds/src/main/java/io/grpc/xds/client/XdsResourceType.java [hunk 2]
```diff
@@ -41,42 +41,38 @@
...
   static final String TYPE_URL_RESOURCE =
       "type.googleapis.com/envoy.service.discovery.v3.Resource";
-  static final String TRANSPORT_SOCKET_NAME_TLS = "envoy.transport_sockets.tls";
+  protected static final String TRANSPORT_SOCKET_NAME_TLS = "envoy.transport_sockets.tls";
   @VisibleForTesting
-  static final String AGGREGATE_CLUSTER_TYPE_NAME = "envoy.clusters.aggregate";
+  public static final String HASH_POLICY_FILTER_STATE_KEY = "io.grpc.channel_id";
   @VisibleForTesting
-  static final String HASH_POLICY_FILTER_STATE_KEY = "io.grpc.channel_id";
+  public static boolean enableRouteLookup = getFlag("GRPC_EXPERIMENTAL_XDS_RLS_LB", true);
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/XdsResourceType.java:48]: We're going to want to grpc flags like this out. We can do that after this move though.

File: xds/src/main/java/io/grpc/xds/XdsClusterResource.java [hunk 3]
```diff
@@ -547,7 +549,7 @@ abstract static class CdsUpdate implements ResourceUpdate {
 
     static Builder forAggregate(String clusterName, List<String> prioritizedClusterNames) {
       checkNotNull(prioritizedClusterNames, "prioritizedClusterNames");
-      return new AutoValue_XdsClusterResource_CdsUpdate.Builder()
+      return new io.grpc.xds.AutoValue_XdsClusterResource_CdsUpdate.Builder()
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsClusterResource.java:552]: Why is `io.grpc.xds` added here?

File: xds/src/main/java/io/grpc/xds/XdsListenerResource.java [hunk 4]
```diff
@@ -51,14 +51,15 @@
 import java.util.Set;
 import javax.annotation.Nullable;
 
-class XdsListenerResource extends XdsResourceType<LdsUpdate> {
+public class XdsListenerResource extends XdsResourceType<LdsUpdate> {
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsListenerResource.java:54]: Remove public?

File: xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java [hunk 5]
```diff
@@ -59,7 +59,8 @@
  * Common base type for XdsClient implementations, which encapsulates the layer abstraction of
  * the xDS RPC stream.
  */
-final class ControlPlaneClient {
+@Internal
+public final class ControlPlaneClient {
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java:63]: It looks like this is public for a single line in GrpcXdsClientImplTestBase. We should just replace the `RpcRetryTask.class.getSimpleName()` with a string literal. (That seems like a code smell that we need to rearrange tests or whatnot, but that replacing it with a string is trivial for right now.)

Ditto for the LoadReportClient tasks.

File: xds/src/main/java/io/grpc/xds/client/ReferenceCounted.java [hunk 6]
```diff
@@ -29,6 +31,7 @@
  * counts being leaked.
  */
 // TODO(chengyuanzhang): move this class into LoadStatsManager2.
+@Internal
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/ReferenceCounted.java:34]: You can remove `@Internal` since it isn't public.

File: xds/src/main/java/io/grpc/xds/client/XdsTransportFactory.java [hunk 7]
```diff
@@ -14,11 +14,12 @@
  * limitations under the License.
  */
 
-package io.grpc.xds;
+package io.grpc.xds.client;
 
 import io.grpc.ExperimentalApi;
 import io.grpc.MethodDescriptor;
 import io.grpc.Status;
+import io.grpc.xds.client.Bootstrapper;
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/XdsTransportFactory.java:22]: Delete import?

File: xds/src/test/java/io/grpc/xds/CdsLoadBalancer2Test.java [hunk 8]
```diff
@@ -783,7 +785,7 @@ public void shutdown() {
     }
   }
 
-  private final class FakeXdsClient extends XdsClient {
+  private static class FakeXdsClient extends XdsClient {
```
Comment (ejona86) [inline xds/src/test/java/io/grpc/xds/CdsLoadBalancer2Test.java:788]: Why no longer final?

File: xds/src/test/java/io/grpc/xds/client/LoadStatsManager2Test.java [hunk 9]
```diff
@@ -14,19 +14,20 @@
  * limitations under the License.
  */
 
-package io.grpc.xds;
+package io.grpc.xds.client;
 
 import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertNotNull;
```
Comment (ejona86) [inline xds/src/test/java/io/grpc/xds/client/LoadStatsManager2Test.java:20]: Use `assertThat(foo).isNotNull()` to keep this test junit-assertions-free.

**Review by ejona86 — COMMENTED** [review]
I've not looked much at XdsClientImpl yet. XdsLogger should be my only non-trivial comment. We'll probably discuss offline to determine what is easiest. (Worst-case, we could inject the class to be used for logging. So this isn't necessarily horrible.)

File: xds/src/test/java/io/grpc/xds/CdsLoadBalancer2Test.java (hunk 8 above)
Comment (ejona86) [reply xds/src/test/java/io/grpc/xds/CdsLoadBalancer2Test.java:788]: How is that related? It can be both static and final. I guess it doesn't matter because this is a test, but still looked strange.

File: xds/src/main/java/io/grpc/xds/BootstrapperImpl.java [hunk 10]
```diff
@@ -17,66 +17,55 @@
...
-  private static final String BOOTSTRAP_CONFIG_SYS_PROPERTY = "io.grpc.xds.bootstrapConfig";
-  @VisibleForTesting
-  static String bootstrapConfigFromSysProp = System.getProperty(BOOTSTRAP_CONFIG_SYS_PROPERTY);
-
-  // Feature-gating environment variables.
-  static boolean enableFederation =
-      Strings.isNullOrEmpty(System.getenv("GRPC_EXPERIMENTAL_XDS_FEDERATION"))
-          || Boolean.parseBoolean(System.getenv("GRPC_EXPERIMENTAL_XDS_FEDERATION"));
+@Internal
+public abstract class BootstrapperImpl extends Bootstrapper {
```
[the remark this answers is not in the document]
Comment (ejona86) [reply xds/src/main/java/io/grpc/xds/BootstrapperImpl.java:46]: This isn't in the client package. Did you intend to move it?

File: xds/src/main/java/io/grpc/xds/GrpcXdsClientImpl.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+import io.grpc.internal.TimeProvider;
+import io.grpc.xds.client.Bootstrapper;
+import io.grpc.xds.client.LoadReportClient;
+import io.grpc.xds.client.XdsClientImpl;
+import io.grpc.xds.client.XdsTransportFactory;
+import io.grpc.xds.internal.security.TlsContextManagerImpl;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+
+public class GrpcXdsClientImpl extends XdsClientImpl {
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/GrpcXdsClientImpl.java:33]: Can we delete this class now, and make `XdsClientImpl` `final`? If we only need this for getServerLrsClientMap(), I'd much rather have the public method just for testing than have `XdsClientImpl` be non-`final` with protected variables.

File: xds/src/main/java/io/grpc/xds/XdsListenerResource.java [hunk 12]
```diff
@@ -51,14 +51,15 @@
...
   static final String ADS_TYPE_URL_LDS =
       "type.googleapis.com/envoy.config.listener.v3.Listener";
   static final String TYPE_URL_HTTP_CONNECTION_MANAGER =
       "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3"
           + ".HttpConnectionManager";
   private static final String TRANSPORT_SOCKET_NAME_TLS = "envoy.transport_sockets.tls";
   private static final XdsListenerResource instance = new XdsListenerResource();
+  protected final FilterRegistry filterRegistry = FilterRegistry.getDefaultRegistry();
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsListenerResource.java:62]: `private`? Ditto in other resource types.

File: xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java [hunk 13]
```diff
@@ -59,6 +59,7 @@
  * Common base type for XdsClient implementations, which encapsulates the layer abstraction of
  * the xDS RPC stream.
  */
+@Internal
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java:62]: You can remove `@Internal` since it is no longer public.

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java [hunk 14]
```diff
@@ -263,12 +234,10 @@ public void run() {
         ImmutableMap.Builder<XdsResourceType<?>, Map<String, ResourceMetadata>> metadataSnapshot =
             ImmutableMap.builder();
         for (XdsResourceType<?> resourceType: resourceSubscribers.keySet()) {
-          ImmutableMap.Builder<String, ResourceMetadata> metadataMap = ImmutableMap.builder();
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:266]: Why was this changed? It looks like previously we returned an immutable map and now we don't. Is something depending on it being mutable now?

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java [hunk 15]
```diff
@@ -88,41 +84,45 @@ public void uncaughtException(Thread t, Throwable e) {
         }
       });
 
-  private final Map<ServerInfo, ControlPlaneClient> serverChannelMap = new HashMap<>();
+  protected final Map<Bootstrapper.ServerInfo, LoadStatsManager2> loadStatsManagerMap =
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:87]: The class is importing `Bootstrapper.ServerInfo`, yet there's a lot of changes from `ServerInfo` to `Bootstrapper.ServerInfo`.

Comment (larry-safran) [reply xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java]: Previously there was an import of Bootstrapper.ServerInfo, so this wasn't really a change.  I've restored the import and gotten rid of the `Bootstrapper` qualifier.

File: xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java [hunk 16]
```diff
@@ -125,11 +127,11 @@ public XdsClient getObject() {
           xdsClient = new XdsClientImpl(
               DEFAULT_XDS_TRANSPORT_FACTORY,
               bootstrapInfo,
-              context,
               scheduler,
               new ExponentialBackoffPolicy.Provider(),
               GrpcUtil.STOPWATCH_SUPPLIER,
-              TimeProvider.SYSTEM_TIME_PROVIDER);
+              TimeProvider.SYSTEM_TIME_PROVIDER,MessagePrinter.INSTANCE,
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java:133]: nit: missing whitespace after the comma. Looks like it should be a newline.

**Review by ejona86 — APPROVED** [review]
